version: '3.8'
services:
  db:
    image: postgres:15-alpine
    restart: always
    environment:
      - POSTGRES_USER=${POSTGRES_USER:?err}
      - POSTGRES_PASSWORD=${POSTGRES_PASSWORD:?err}
      - POSTGRES_DB=${POSTGRES_DB:?err}
    healthcheck:
      test: pg_isready -q -d $POSTGRES_DB -U $POSTGRES_USER
      interval: 5s
      timeout: 30s
      retries: 3
    
  server: 
    build: ./server/app
    volumes:
      - ./server/app:/server
    environment:
      - DB_USER=${POSTGRES_USER:?err}
      - DB_HOST=${POSTGRES_HOST:?err}
      - DB_PASSWORD=${POSTGRES_PASSWORD:?err}
      - DB_PORT=${POSTGRES_PORT}
      - DB_DATABASE=${POSTGRES_DB:?err}
    ports:
      - '80:80'
    depends_on:
      db:
        condition: service_healthy
  
  client:
    build: ./client

    volumes:
      - ./client:/app
      - /app/node_modules
    ports:
      - '3000:3000'
    
    depends_on:
      - server

  npm:
    image: node:18-alpine
    working_dir: /app
    volumes:
      - ./client:/app

    ports:
      - '3000:3000'
    stdin_open: true
    tty: true
  
